### PR TITLE
Support nested aggregations

### PR DESCRIPTION
**Pull request by nfantone**
@danpaz Consider this a draft, for now - but I believe it's looking quite decent. I've tried to modify as little as I could of what there already was, so I opted for a composition pattern for the internal implementation.

You can now nest aggregations with the following API:

```javascript
const body = new Bodybuilder()
  .filter('terms', 'chain.code', chain)
  .filter('range', 'timestamp', { gte: new Date, lte: new Date })
  .orFilter('terms', 'region.code', ['LOC', 'LAC', 'LIC'])
  .aggregation('terms', 'customer.name', 'byName',
    agg =>
      agg.aggregation('terms', 'bis')
        .aggregation('max', 'bor')
        .aggregation('min', 'bas',
          agg2 => agg2.aggregation('missing', 'foo'))
  ).build()
```

This snippet showcases the main points of the new functionality:

- Keep the builder design pattern going. Mixing of filters/queries and nested aggregations is supported.
- Allow defining an optional nesting-`function` as the `.aggregation` last parameter. The `function` receives the recently created aggregation and is expected to return some new object that will, ultimately, become the actual body of the aggregation.
- Aggregation instances are now builders of their own. You can call `.aggregation` on them, exactly as you would on the main `BodyBuilder`. Calls are chainable, as expected.
- Multiple-level of nesting is also supported.

I've added units tests for this, as well.

Please, let me know your thoughts!

### CODE REVIEW

**Comment by nfantone** [comment]
Ok, so it seems to be failing on older versions of node due to the lack of support for `Object.setPrototypeOf` and `Object.freeze`.

First one may be avoided by using `Object.create` + mixins (like `_.extend`) or classic inheritance. The latter, is really just a cautionary _nice-to-have_ and it could be removed entirely.

**Comment by danpaz** [comment]
Nice!!! I'm really excited about this PR! The pattern looks solid to me, and I think could probably be extended to work equally well with nested filters and nested queries. 

I'll take a closer look in the next few days. Regarding `Object.freeze`, I think you might get the functionality you want from `_.clone`.

**Comment by nfantone** [comment]
@danpaz Yes, `_.clone` _may_ work (sort of), but I was looking to avoid duplication that could potentially generate a new object that'd garbage collected just instants after if was requested. Moreover, unless I `_.cloneDeep` it (which is expensive), the copy still holds references to modifiable elements, like arrays.

We could always return the internal aggregation object and document the exposition, warning users to be cautious about it.

**Comment by nfantone** [comment]
@danpaz There. Checks are passing now.

File: src/aggregations/aggregation-builder.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import _ from 'lodash'
+import aggregations from './index'
+
+const AggregationBuilder = {
```
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: Not critical but why did you chose a plain old object instead of a es2015 class? `Object.create(X)` seems like a non-standard alternative to `new X()` but maybe I'm missing something.

File: src/aggregations/aggregation-builder.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    // If last argument is a nesting function, remove it
+    // from `args` before building the aggregation
+    let nest = _.last(args)
+    args = _.isFunction(nest) ? _.initial(args) : args
+
+    // Mixin to assign newly built aggregation properties
+    // to the `AggregationBuilder`
+    _.extend(aggregation, klass(...args))
```
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: IIRC `_.extend` and `_.assign` are aliases of each other, so let's stick to just one.

File: src/aggregations/aggregation-builder.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    // Extend the current aggregation object with the
+    // recently built aggregation
+    this._aggs = _.assign(this._aggs, aggregation)
+
+    if (_.isFunction(nest)) {
+      // Resolve the nested aggregation and set it as a
+      // children of the current aggregation
```
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: typo `child`

File: src/aggregations/aggregation-builder.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    return this
+  },
+
+  /**
+   * Alias for `aggregation`.
+   *
+   * @param {...[Object]} args Arguments for `aggregation`.
+   * @returns {AggregationBuilder} Builder object.
+   */
+  add(...args) {
```
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: Why is this needed?

**Comment by danpaz** [comment]
@nfantone Looking really good! Just a few small comments, back to you.

File: src/aggregations/aggregation-builder.js (hunk 2 above)
Comment (nfantone) [inline src/aggregations/aggregation-builder.js]: No, they are not. `_.extend` is an alias for `_.assignIn`, which iterates over inherited properties as well. But I'll change it to `_.assignIn` if you prefer that.

File: src/aggregations/aggregation-builder.js (hunk 1 above)
Comment (nfantone) [inline src/aggregations/aggregation-builder.js]: Many reasons. First one being that `Object.create` is _not_ an alternative to `new`. If anything, you could achieve the same using something like:

```javascript
function X() {}
var o = Object.create(X.prototype) // This is essentially new X()
```
Javascript has no classes, despite whatever the literature or SO questions may love to express. The `class` syntax is (almost) just syntactic sugar on top of the classic `[[Prototype]]` mechanism. 

And here, we don't need all that sugar. We just need to _compose_ objects. Aggregations modules need to "become builders". The alternative would be to create a class and have all your `*-aggregation.js` functions become classes that extended the builder, thus modifying 30% of the files in the project.

I'd _highly_ recommend reading YDKJS for more on composition and javascript class illness: https://github.com/getify/You-Dont-Know-JS/blob/master/this%20&%20object%20prototypes/ch6.md

File: src/aggregations/aggregation-builder.js (hunk 4 above)
Comment (nfantone) [inline src/aggregations/aggregation-builder.js]: It's not, really. But it makes for a nicer syntax when using the builder inside the main `BodyBuilder` class: https://github.com/danpaz/bodybuilder/pull/16/files#diff-1fdf421c05c1140f6d71444ea2b27638R270

Shall I remove it?

**Comment by nfantone** [comment]
@danpaz Reviewed and commented on them!

File: src/aggregations/aggregation-builder.js (hunk 2 above)
Comment (nfantone) [inline src/aggregations/aggregation-builder.js]: Turns out, we could get away with just plain ol' `_.assign` here, so I changed it to that.

File: src/aggregations/aggregation-builder.js (hunk 3 above)
Comment (nfantone) [inline src/aggregations/aggregation-builder.js]: Fixed!

File: src/aggregations/aggregation-builder.js (hunk 4 above)
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: Gotcha, thanks, I'm fine keeping it.

File: src/aggregations/aggregation-builder.js (hunk 1 above)
Comment (danpaz) [inline src/aggregations/aggregation-builder.js]: I understand `class` is just sugar for prototypal inheritance and that composition over inheritance can be cleaner, but with what's implemented here there wouldn't be much difference either way. But this already works so instead of getting into a philosophical debate about JS inheritance I'm fine leaving this as is :)

**Comment by danpaz** [comment]
LGTM. Just merge conflicts now 😸

**Comment by npatmaja** [comment]
👍   awesome, +1 for this PR

**Comment by danpaz** [comment]
@nfantone can you also update the readme with instructions on how to use this new feature?

**Comment by nfantone** [comment]
@danpaz Will do!

**Comment by nfantone** [comment]
@danpaz Done with docs. Anything else?

**Comment by danpaz** [comment]
@nfantone Docs look good! Could you merge in the upstream changes and resolve merge conflicts?

**Comment by nfantone** [comment]
@danpaz That should do it. Conflicts came from the browser build. I take it it's not something usual, but you'll _always_ have conflicts on `bodybuilder.min.js` when there are multiple pending PRs. You should consider distributing things through [bower](https://bower.io) and npm and move the [deploy to Travis](https://docs.travis-ci.com/user/deployment/npm/).

**Comment by danpaz** [comment]
Cool! I didn't know travis had the ability to publish packages. I'll try that out. 

Thanks for your A++ efforts on these features, it is very much appreciated! 🎉
